File: src.lib/field/kit/particle/Particle.scala
/*                                                                            *\
**           _____  __  _____  __     ____     FieldKit                       **
**          / ___/ / / /____/ / /    /    \    (c) 2009, field                **
**         / ___/ /_/ /____/ / /__  /  /  /    http://www.field.io            **
**        /_/        /____/ /____/ /_____/                                    **
\*                                                                            */
/* created March 31, 2009 */
package field.kit.particle

import field.kit.Logger

/** a single particle from the flock */
class Particle extends Logger {
  import field.kit.math.Vec3
  fine("init")
  
  // set when particle is being added to a flock
  var flock:Flock[_] = null
  var ps:ParticleSystem = null
  
  var age = 0f
  var lifeTime = 10 * 1000f
  
  var position = new Vec3
  var velocity = new Vec3
  var steer = new Vec3
  var steerMax = 1f
  var velocityMax = 10f
  
  protected val absVelocity = new Vec3

  def update(dt:Float) {
    age += dt
    
    // update driving force
    steer.clamp(steerMax)
    velocity += steer
    velocity.clamp(velocityMax)
  
    // make velocity time invariant
    absVelocity(velocity) *= (dt / ps.timeStep)
    position += absVelocity

    // clean up
    velocity *= ps.friction
    steer.zero
  }
}

File: FieldKit.js/src/particle/Particle.js
/*!
 * FieldKit JavaScript Library
 * http://code.google.com/p/fieldkit
 *
 * Copyright (c) 2010 Marcus Wendt
 * Licensed under the LGPL license.
 * http://www.gnu.org/licenses/lgpl.html
 *
 * Created: June 20, 2009
 */


/**
 * 2D Particle
 */
fk.particle.Particle = fk.Class.extend({
	
	init: function(flock) {
		this.flock = flock;

		this.position = new fk.math.Vec2();
		this.velocity = new fk.math.Vec2();
		this.steer = new fk.math.Vec2();
		
		this.age = 0;
		this.steerMax = 1.0;
		this.velocityMax = 10;
		this.rotation = 0;
		this.turningSpeed = 0.1;
		this.size = 10;

		// temp fields
		this.theta = 0;
		this.absVelocity = new fk.math.Vec2();
				
		// reset to defaults
		this.reset();
	},

	reset: function() {
		this.velocity.zero();
		this.steer.zero();
		
		this.age = 0;
		this.rotation = 0;
		this.size = 10;
		
		this.theta = 0;
		this.absVelocity.zero();
	},

	updateLogic: function(dt) {
		this.age += dt;
		this.steer.clamp(this.steerMax);
		this.velocity.addV(this.steer).clamp(this.velocityMax);
		this.steer.zero();
		this.theta = -Math.atan2(this.velocity.x, this.velocity.y);
	},
		
	updateSimulation: function(dt) {
		this.absVelocity.setV(this.velocity).mulS(dt / this.flock.ps.updateTicks);
		this.position.addV(this.absVelocity);
		this.velocity.mulS(this.flock.ps.friction);
		this.rotation = fk.math.slerpAngle(this.rotation, this.theta, this.turningSpeed);
	},
});
